DNS: improve the responsiveness of DNS editor

klimeryk opened the pull request:
The UX for the DNS editor could use some improvements so that it's more responsive.

- [x] Switch to optimistic deletion - remove the record from the view immediately, and from store after confirmation from the backend.
- [x] Disable Add button when submitting a new DNS record

Testing:
 * deletion should be instantaneous, that is the DNS record should be removed from the list immediately upon clicking the remove button. However, only when a successful response is received from the backend that the record is removed from the store on the frontend. If an error is received, my "undelete" the record, show it again to the user and display an error notice.
 * when adding a new record, the Add and Cancel buttons should be disabled. After the form has been submitted (successfully or not), they should be re-enabled.

/cc @umurkontaci @aidvu (have at it, guys ;P)
Thanks to @dzver for the original report and the idea to improve the DNS editor's responsiveness.

stephanethomas commented on client/my-sites/upgrades/domain-management/dns/dns-add-new.jsx:
```diff
@@ -33,7 +33,8 @@ const DnsAddNew = React.createClass( {
 			show: false,
-			fields: null
+			fields: null,
+			submittingForm: false,
```
What about moving this state [in the store](https://github.com/Automattic/wp-calypso/blob/master/client/lib/domains/site-redirect/reducer.js#L24) instead? That's what we've been doing for some time now and it works [pretty well](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/upgrades/domain-management/site-redirect/index.jsx#L111). It helps to centralize state management and to reduce the complexity of components.

klimeryk replied:
Very good suggestion! I must start thinking more with ~~portals~~ stores...

stephanethomas commented on client/my-sites/upgrades/domain-management/dns/dns-list.jsx:
```diff
@@ -38,7 +34,7 @@ const DnsList = React.createClass( {
 					key={ index }
 					dnsRecord={ dnsRecord }
-					deleteDns={ this.deleteDns }
+					deleteDns={ this.handleDeleteDns }
```
I think you should keep the former name for this method. That's the [convention](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#react-method-names) we've been following so far. Or maybe I'm missing something :).

klimeryk replied:
I was following @gziolo's suggestion from `p4TIVU-1GQ-p2`:
> Prefix methods that triggers action, that sends data to the server or changes state on the page, with `handle`.

I thought it was a nice suggestion, but I'm open to discussion - maybe this convention did not take or there's an even cooler prefix on the block (like `dealWith` or `takeCareOf` :stuck_out_tongue_winking_eye: ).

gziolo replied:
@stephanethomas is right. There was long discussion how to name methods. In the end we decided to use method names that better reveal intention :) So, yes, we should keep `deleteDns`.

klimeryk replied:
My bad then - I'll revert it to `deleteDns` :+1:

stephanethomas commented on client/my-sites/upgrades/domain-management/dns/dns-record.jsx:
```diff
@@ -91,6 +91,10 @@ var DnsRecord = React.createClass( {
 	},
 
 	render: function() {
+		const isBeingDeleted = this.props.dnsRecord.isBeingDeleted && this.props.dnsRecord.isBeingDeleted === true;
```
We could probably use following shorter approach here:

```javascript
if ( this.props.dnsRecord.isBeingDeleted ) {
  return null;
}
```

klimeryk replied:
Heh, you're right, I was trying to be extra careful, but I got too carried away :grin:

umurkontaci commented on client/lib/domains/dns/reducer.js:
```diff
@@ -46,6 +44,36 @@ function deleteDns( state, domainName, record ) {
+	return update( state, command );
+}
+
+function findDns( state, domainName, record ) {
```
`findDns` is probably not the right name for a function that returns index. How about `findDnsIndex`?

klimeryk replied:
:+1:

umurkontaci commented on client/lib/domains/dns/reducer.js:
```diff
@@ -46,6 +44,36 @@ function deleteDns( state, domainName, record ) {
+
+	return findIndex( state[ domainName ].records, matchingFields );
+}
+
```
It seems you don't need `state` or `domainName` in this function. You can just pass in `records`.

    function findDnsIndex( records, { id, data, name, type } ) {
      const fields = isUndefined( id ) ? { data, name, type } : { id, data, name, type };
      return findIndex( records, matchingFields );
    }

klimeryk replied:
Good catch! I just extracted this method and did not notice that the arguments could be simplified.

umurkontaci commented on client/lib/domains/dns/reducer.js:
```diff
@@ -46,6 +44,36 @@ function deleteDns( state, domainName, record ) {
 }
 
+function markDnsForDeletion( state, domainName, record, isBeingDeleted ) {
+	const index = findDns( state, domainName, record ),
```
I know it's convenient, but I think it'd be better to use an object to pass extra details (`isBeingDeleted` in this instance), so that when you're calling the function I know that all the parameters mean without checking to the original function's signature.

klimeryk replied:
I was torn on this - I agree that it's more readable, but on the other hand it's less "strict" - you can pass any object, without the `isBeingDeleted` property. On the _other_ hand, it's an internal function, so that shouldn't be an issue (not sure if it's at all :stuck_out_tongue:), so I'll go with your suggestion :+1:

umurkontaci commented:
Works as promised, but as I'm deleting something I felt like something is broken. I had to refresh to confirm that it was deleted. We should at least add a notice to let the user know that they've successfully deleted something or maybe instead of hiding the item completely, we can make it look like "disabled".

They can proceed to other actions while one of the records is in the disabled state. 

It is out of scope but if we add a "disabled" state we can actually add a "undo" function. Removing a DNS record is potentially destructive and it's be nice to have a confirmation dialog or "undo" button to keep things in order.

I would really hate if I had deleted my records while trying to scroll on my phone to add a record.

klimeryk commented:
@umurkontaci: thanks for the suggestion, I had the same feeling as you, but didn't know how to address it :grin:
The changes I've added fade out the record being deleted (and, of course, don't allow deleting the same record again) and remove it after getting confirmation from the server.

As for the undo suggestion - I agree that it'd be best to approach it like that. I've read somewhere that it's a better flow - don't obstruct the user with confirmation modals, but offer easy undo instead. I've opted for an easier implementation (one that requires no backend modifications). So, we delete the record like we did previously, but in the success notice I've added an "Undo" action that basically calls the `addDns` action to add the same record that was just deleted. Ideally, we would mark the record as "delete unless cancelled in the next 10 seconds", but this should be fine too.

As a result of these changes I also revised our form for adding new records and made a simple change from:
![2016-01-14-005709_669x153_scrot](https://cloud.githubusercontent.com/assets/3392497/12311974/a1b52a2c-ba5b-11e5-984c-3ac434587d7d.png)
to:
![2016-01-14-005556_757x140_scrot](https://cloud.githubusercontent.com/assets/3392497/12311953/760dc3e8-ba5b-11e5-9e9e-bea247ed784b.png)

I think using the form input with a suffix conveys better our intention - the user should just input the the subdomain part now. And just to be on the safe side, I also support if he ignores us and instead of `test`, inputs `test.example.com`. In both cases, the saved record will be `test.example.com`.

mahangu commented:
Just did some testing, and I was able to add a CNAME record that started with a `-`, such as `-example`. According to the RFC, subdomains can only begin and end with alphanumeric characters, and therefore this URL returns a 400 Bad Request.

Steps to reproduce:
1) Create a new CNAME record with `-example` in the Host field.

I also tested ending with a hyphen at the end of the subdomain, like `example-`, but for that we correctly return the error message 'Invalid Domain or Subdomain' when you click 'Add New DNS Record'. However, we still do not display the 'Invalid Name' warning message in real time when the Host field is edited.

klimeryk commented:
Switching to `[Status] In Progress` while I fix up @mahangu's findings and the other GitHub issues I've linked to this one.

BandonRandon commented:
Just tested this branch locally and it does fix some of the issues mentioned in: https://github.com/Automattic/wp-calypso/issues/1604#issuecomment-173687235

I would still like to be able to strip spaces at the front or end. I also really like how you added the domain name label to the end. However, it would be good to still check for `blog.example.com` in the event the user doesn’t want `blog.example.com.example.com` but often users miss these details. 

My only other comment is when the record fails showing the `Invalid Domain or Subdomain` error I'm taken back to the 'A' record in the drop down instead of the 'CNAME' I was trying to create.

klimeryk commented:
Thank you @BandonRandon for testing :)
As for inputting `blog.example.com` it should work as you would expect - `blog.example.com` will be added instead of `blog.example.com.example.com`. Like you, I also assumed that some users will write the whole domain and added an special case for that :grin: Have you tested this behaviour on this branch and it's not working?

Thanks for the other comments - I'll be looking into them, improving this PR and will let you know when it can be tested again.

BandonRandon commented:
> As for inputting blog.example.com it should work as you would expect - blog.example.co
It looks like adding `blog.example.com` does work, as you mentioned. 

What I was testing and failed was `blog.EXAMPLE.COM` This lets you submit then comes back with a "Invalid Domain or Subdomain" message.

klimeryk commented:
After a chat with <del>Master Yoda</del>@dzver I've decided to wrap up the PR at this stage. There are some validation issues left (including the ones reported by @mahangu), but I'll be opening a new PR for those. Otherwise it'd be hard to review this monstrosity of a PR.
I believe the main goal of this PR, improving the responsiveness of the editor, has been reached and I ask you to test that aspect mostly. If you find issues with validation (that have not been introduced by this PR), please check the existing issues and report if needed. Thank you for your cooperation :smile: 

/cc @aidvu (in case you wanted to look at some fugly Calypso code :grin:)

umurkontaci commented on client/lib/domains/dns/index.js:
```diff
@@ -86,32 +75,44 @@ function isValidData( data, type ) {
+	return normalizedRecord;
+}
+
+function getNormalizedName( name, type, selectedDomainName ) {
```
Can you add some docs about what these functions do, or rename the functions to describe what they're doing?

klimeryk replied:
I will be refactoring this whole file in a separate PR, so for now I'd prefer to leave it as-is (the names are nondescriptive, that's for sure), if that's OK with you.

umurkontaci replied:
:+1:

umurkontaci commented on client/lib/domains/dns/index.js:
```diff
@@ -86,32 +75,44 @@ function isValidData( data, type ) {
-function removeTrailingDomain( domain, trailing ) {
-	return domain.replace( new RegExp( '\\.+' + trailing + '\\.?$', 'i' ), '' );
+function isARecord( type ) {
+	return includes( [ 'A', 'AAAA' ], type );
```
AAAA record is not an A record

klimeryk replied:
Heh, it was a mental shortcut, but you are right, the function's name could be misleading :+1:

umurkontaci commented on client/my-sites/upgrades/domain-management/dns/dns-record.jsx:
```diff
@@ -1,7 +1,9 @@
-var React = require( 'react' );
+import React from 'react';
+import classNames from 'classnames';
+import endsWith from 'lodash/string/endsWith';
```
No need for lodash here: Use https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/String/endsWith

klimeryk replied:
Actually, yes, I believe we should use lodash here. There's been problems with `endsWith` [in the past](https://github.com/Automattic/wp-calypso/pull/1777) (in related news, `startsWith` recently caused [quite a ruckus](https://calypsop2.wordpress.com/2016/01/13/calypso-was-broken-in-ie11safari8-since-yesterday/) too) because it's not polyfilled (last time I checked), so it will break on older browsers.

umurkontaci replied:
:+1: Didn't know we weren't polyfilling it.

umurkontaci commented on client/my-sites/upgrades/domain-management/dns/dns-record.jsx:
```diff
@@ -75,24 +77,45 @@ var DnsRecord = React.createClass( {
-		return isRoot ? domain : name + '.' + domain;
+		if ( endsWith( name, '.' ) ) {
+			return name.slice( 0, -1 );
+		}
```
Is it possible to use the same approach in the normalize function above? I feel like it'd make it much simpler.

klimeryk replied:
Not sure which normalize function you mean here - `getNormalizedName` already uses `endsWith`.

umurkontaci replied:
Maybe I'm missing something but I think `getNormalizedName` can be simplified without creating a dynamic regex. You can create another PR for this

umurkontaci commented:
Couple of things,
- It seems like we're not doing error handling. I've cut my connection and tried to add something, got this:

![image](https://cloud.githubusercontent.com/assets/991022/12665997/3e2f1dd6-c5f1-11e5-8670-a811fd04163d.png)

- We should probably add a success notice when you add something.

- We should disable the buttons when it's in progress.

- I was able to set A record and CNAME recording to the same point. I shouldn't be able to do that.

- I was annoyed of the form fields moving due to something being added to top. The same thing applies for the inline form errors, they move the fields down. 

- Did you get a design review?

umurkontaci commented:
PS: If you're making changes to the code from now on, please don't rebase until you got a green light since it's already big and will be hard to track what changed.

stephanethomas commented:
> PS: If you're making changes to the code from now on, please don't rebase until you got a green light since it's already big and will be hard to track what changed.

I'm not sure to understand, could you clarify that when you get the chance @umurkontaci? As far as I know the only downside of rebasing is that GitHub collapses inline comments. On the other hand, rebasing often is a good way to spot as soon as possible any issue that could crop up from integrating these changes with `master`.

klimeryk commented:
> Did you get a design review?

@breezyskies, pretty please? :grin: With cherry on top! :cherries:

breezyskies commented:
Looks good overall, nice improvements! A few small nitpicks:

The cancel button should probably be below the primary button in mobile, and have a margin (looks like this predated your PR, so feel free to open a separate issue):

![screen shot 2016-01-29 at 5 30 46 am](https://cloud.githubusercontent.com/assets/3011211/12676864/ff658cd0-c649-11e5-8e17-f854bc8bc9ae.png)

+1 for the undo option, but I might sync this up with @mikeshelton1503's design in https://github.com/Automattic/wp-calypso/issues/2531. (We intentionally decided against including the dismiss, since the undo is an important link there, and it will disappear on refresh or when the user navigates away.)

umurkontaci commented:
> I'm not sure to understand, could you clarify that when you get the chance @umurkontaci? As far as I know the only downside of rebasing is that GitHub collapses inline comments. On the other hand, rebasing often is a good way to spot as soon as possible any issue that could crop up from integrating these changes with master.

This is huge PR and it would be really hard for reviewers to understand what happened since their last review if their request gets buried into the rest of the changes. Nothing wrong with rebases in general, but they do make the reviews harder when the PR itself is big; I am suggesting holding on the rebasing until the changes in the code is complete.

klimeryk commented:
> It seems like we're not doing error handling. I've cut my connection and tried to add something, got this [...]

Good catch, I got that empty error notice some time ago and couldn't reproduce then. Now it's fixed :+1: 

> We should probably add a success notice when you add something.

Originally, there was a success notice, when a DNS record got added. But it seemed too verbose with the new approach. My "vision" here is that the user inputs the DNS details, hits Add and the form is immediately available for adding another record. The new record is added at the end of the list of records, just above the form, so the user should spot it. When the record is being added, it's grayed out and disabled. When it's been successfully added (the server responded with an A-OK), it gets enabled. Seems pretty intuitive to me - slapping a huge success notice on top of it seemed superfluous to me. But I'd love a second opinion on this - @breezyskies? :smiley: I am looking at this from a power user point of view, so I'm biased towards a more quick, optimised flow (which might not be ideal for everyone).

> We should disable the buttons when it's in progress.

As mentioned in the previous point, there's no reason to disable the buttons when it's in progress - you can safely add another record when the previous one is in progress.

> I was able to set A record and CNAME recording to the same point. I shouldn't be able to do that.

True, but it's beyond the scope of this PR. I've created a separate issue for this: #2920

> I was annoyed of the form fields moving due to something being added to top. The same thing applies for the inline form errors, they move the fields down.

That's a standard "feature" of all forms in Calypso, from what I've seen. But your comment made me go over the forms again (so thank you for the nudge :)) and I made them more uniform with other forms in Calypso:
 * I've put the validation error below the field it is referring to (before it was above it)
 * when the field has a validation error, I mark it appropriately using the `isError` property (which styles the field appropriately)
@breezyskies, can you perhaps weight in here? Can we do something about this "jumping"? Is this a bug or a "feature"? :grin:

Thanks for all the comments and code review, @umurkontaci - keep 'em coming :grinning:

umurkontaci commented:
> Originally, there was a success notice, when a DNS record got added. But it seemed too verbose with the new approach. My "vision" here is that the user inputs the DNS details, hits Add and the form is immediately available for adding another record. The new record is added at the end of the list of records, just above the form, so the user should spot it. When the record is being added, it's grayed out and disabled. When it's been successfully added (the server responded with an A-OK), it gets enabled. Seems pretty intuitive to me - slapping a huge success notice on top of it seemed superfluous to me. But I'd love a second opinion on this - @breezyskies? :smiley: I am looking at this from a power user point of view, so I'm biased towards a more quick, optimised flow (which might not be ideal for everyone).

I would disagree on two levels. First off, the user might not have a full view (small screen) or maybe not looking the list and might miss the addition. Adding a success message does not disrupt or slow down the user, I think. Besides, as a power user, I don't trust anything to work on web. So if something acts like working immediately (0ms wait), I don't trust it, I wait for another confirmation (graying out helps) to tell me it actually worked. Otherwise I do a full refresh to ensure it worked.

> As mentioned in the previous point, there's no reason to disable the buttons when it's in progress - you can safely add another record when the previous one is in progress.

Technically, yes they should be able to. But we're disabling buttons throughout Calypso (the element of least surprise), combined with above, I still think we'd need to add a disabled state to buttons. It's not like users are going to add tons of records very fast that the buttons would slow them down. @breezyskies might weigh in here.

umurkontaci commented:
Works as expected. :+1:

breezyskies commented:
> The new record is added at the end of the list of records, just above the form, so the user should spot it. When the record is being added, it's grayed out and disabled. When it's been successfully added (the server responded with an A-OK), it gets enabled.

The grayed out pattern is nice, but I agree a success message wouldn't hurt here. I'm not sure how relevant the "power DNS user" is to our user base (probably mostly relevant to us as _testers_, not even as "real" usage). DNS is arguably one of the most complicated and technical features of WordPress.com, so over-communicating is probably the way to go.

> It's not like users are going to add tons of records very fast that the buttons would slow them down.

Agree with this one, it's more consistent with the rest of Calypso and (I think) helps reinforce the action in progress. 

> I've put the validation error below the field it is referring to (before it was above it)

There is an [existing component](https://wpcalypso.wordpress.com/devdocs/design/form-fields) for inline errors, just in case you hadn't seen it yet.

> when the field has a validation error, I mark it appropriately using the isError property (which styles the field appropriately) @breezyskies, can you perhaps weight in here? Can we do something about this "jumping"? Is this a bug or a "feature"?

I wouldn't call it either a bug or a feature, just something to improve on at the component level (maybe through subtle animation). I would say this is beyond the scope of this PR. :)

klimeryk commented:
> The grayed out pattern is nice, but I agree a success message wouldn't hurt here. I'm not sure how relevant the "power DNS user" is to our user base (probably mostly relevant to us as testers, not even as "real" usage). DNS is arguably one of the most complicated and technical features of WordPress.com, so over-communicating is probably the way to go.

and

> Agree with this one, it's more consistent with the rest of Calypso and (I think) helps reinforce the action in progress.

Well, can't argue with another user of an [X100S](http://www.fujifilm.com/products/digital_cameras/x/fujifilm_x100s/) :grin: Done and done :+1: 

> There is an existing component for inline errors, just in case you hadn't seen it yet.

Yup, that's what was being used here from the start - it was just placed before the field, not after, for some reason.

@umur:
> This is huge PR and it would be really hard for reviewers to understand what happened since their last review if their request gets buried into the rest of the changes. Nothing wrong with rebases in general, but they do make the reviews harder when the PR itself is big; I am suggesting holding on the rebasing until the changes in the code is complete.

I'm not quite sure I follow you. If there are conflicts with `master`, a rebase could introduce some changes to the code under review, but of course the rebase is needed ASAP. On the other hand, if there are no conflicts (like it was during this PR, IIRC), rebasing shouldn't change the code being added as part of the PR (just the commits' hashes) and thus your comments should be still visible and valid (not shown as "outdated"). Please correct me if I'm wrong - but maybe let's move the conversation to Slack if that's the case, because this is not the best place and I don't want to derail this PR :laughing: 

Anyway, thanks again for the comments - I think I've fixed all the issues that you wonderful lot raised :bow:

umurkontaci commented:
> I'm not quite sure I follow you. If there are conflicts with master, a rebase could introduce some changes to the code under review, but of course the rebase is needed ASAP. On the other hand, if there are no conflicts (like it was during this PR, IIRC), rebasing shouldn't change the code being added as part of the PR (just the commits' hashes) and thus your comments should be still visible and valid (not shown as "outdated"). Please correct me if I'm wrong - but maybe let's move the conversation to Slack if that's the case, because this is not the best place and I don't want to derail this PR :laughing:

When the PR gets really big, it is hard to digest all the changes at once. When the reviewer finally digests it, they can do the subsequent reviewers incrementally (take from the last reviewed commit). But if you do change the history through rebase, by squashing commits etc. You're forcing the reviewer the review the whole code change again since it is now impossible to track what has been previously reviewed.

Maybe I wasn't clear in the beginning, rebasing with master if you have conflicts is alright, squashing, splitting commits is not. I think this only applies when the PR is big, doesn't matter that much to small PRs and you can always rebase / squash before merging.

In fact, I think this needs a history cleanup before the merge, (you don't need a different commit for a refactor of the code you introduced in the same branch).

BandonRandon commented:
> When the PR gets really big, it is hard to digest all the changes at once. '

Thanks for bringing this up @umurkontaci. As you may know I've been following this issue for a while and trying to test it before we launch it to our users. I really do love the changes that this will bring but I've been finding it difficult to test and follow each of the things this PR sets out to address. 

For example, there are several issues linked to this one PR, and I'm not sure if I should test all of them and leave notes or just the most recently mentioned. 

@klimeryk can you confirm that this sets out to fix all of the following:
![screenshot_02-02-2016-20 46 00](https://cloud.githubusercontent.com/assets/150348/12773230/4633fda2-c9ee-11e5-8e23-d38ca1ee017b.png)

aidvu commented:
Works for me. Buttons disabled on add. Not disabled on delete, but disabled on Undo. A thing I dislike is the persistent messaging on top.

Also check the errors in the console. My guess is that the errors are generated with the same IDs.

![test dns editor](https://cloud.githubusercontent.com/assets/1103398/12783587/b79e878e-ca82-11e5-8112-155af5cc603a.png)

aidvu commented:
And found a bug when trying to add the same thing twice.

The endpoint returns success, which is wrong, and you can see that the list of ids is undefined.

![dns editor duplicate bug](https://cloud.githubusercontent.com/assets/1103398/12783814/e5d1ba12-ca83-11e5-8c0b-3770ceaa5bef.png)

klimeryk commented:
Thank you, @aidvu for testing! The bugs you found weren't caused by this PR, but I've submitted PRs to fix them (#3063 and #3064), since it turned out it some simple changes were needed only. The notices mechanism is in flux atm, hence the rough edges here and there.

klimeryk commented:
@umurkontaci & @BandonRandon: as mentioned in my [previous comment](https://github.com/Automattic/wp-calypso/pull/2262#issuecomment-175691993), I should have focused on _just_ improving the responsiveness of the DNS editor in this PR. But while working on that, I had to change some other aspects of the editor - and in result this PR got bloated.
I've learned my lesson and will be more careful in the future, but right now I'd like to finally merge this in :persevere: 
So please:
> There are some validation issues left (including the ones reported by @mahangu), but I'll be opening a new PR for those. Otherwise it'd be hard to review this monstrosity of a PR.
I believe the main goal of this PR, improving the responsiveness of the editor, has been reached and I ask you to test that aspect mostly. If you find issues with validation (that have not been introduced by this PR), please check the existing issues and report if needed. Thank you for your cooperation :smile:

Unfortunately, GitHub doesn't seem to remove references to issues, even if I remove the reference from the other side... (or maybe it's cached and it will update eventually). So please ignore the linked issues ~_~

@umurkontaci: I've rebased, redone some commits and removed unnecessary commits from the history, as requested.

BandonRandon commented:
Just tested responsiveness. The only thing I noticed is there seems to be some missing padding in the save buttons:
![screenshot_03-02-2016-23 43 17](https://cloud.githubusercontent.com/assets/150348/12809011/666c6a04-cad0-11e5-85b7-de9179f78b84.png)

Here is what it looks like with some margin:
![screenshot_03-02-2016-23 50 48](https://cloud.githubusercontent.com/assets/150348/12809089/0adaece6-cad1-11e5-98fa-60841c995856.png)


I also looked though this breakpoint in other pages and could not find anywhere else we have two buttons making me think it may be a good idea to offer a different way to cancel or save.

klimeryk commented:
@BandonRandon: sorry if I confused you (argh, again :persevere:) - by "responsiveness" I meant, that the editor should be more responsive to the user's actions. So it should provide instant feedback, etc. (see the first comment in this PR). Not in the sense that it should be mobile friendly, etc :laughing: 
Actually, @breezyskies already [mentioned this issue](https://github.com/Automattic/wp-calypso/pull/2262#issuecomment-176759355) in this PR and it so happens I've provided a fix for it in #3026. Feel free to test it there :) (see, I _can_ learn from my mistakes :wink:)

BandonRandon commented:
> I meant, that the editor should be more responsive to the user's actions.
In this case the PR seemed to do that job nicely. :)

umurkontaci commented:
There's something off with the notices. Also this has a conflict with master.

![image](https://cloud.githubusercontent.com/assets/991022/12862576/b7191a40-cc21-11e5-9847-50c5b7a66302.png)

Once the notification issue is fixed and branch is rebased with master, this PR is good to go. Anything not a blocker can be handled in another PR.

klimeryk commented:
Thanks for looking at this again, @umurkontaci. The conflict was introduced today by a small change in #3104 - it's been fixed now, thanks for bringing that to my attention (I wish GitHub would notify about such things ;/).
About the notice - I've noticed (ha-ha) that too, but it's not a problem introduced by this PR, but with the new global-notices that have rounded corners. I should have reported it sooner... but it looks like it's been fixed in #3122 in the meantime :smiley: The rebase I've just performed pulled in that change and it looks OK now.

umurkontaci commented:
@klimeryk Looks good! This is good to merge :)

I only have one small feedback which should be addressed later: When the user clicks undo, we need to show a notice.
